plugins/civmodcore-paper/src/main/java/vg/civcraft/mc/civmodcore/inventory/items/updater/impl/CustomItemsUpdater.java

```diff
@@ -0,0 +1,102 @@
+package vg.civcraft.mc.civmodcore.inventory.items.updater.impl;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.bukkit.NamespacedKey;
+import org.bukkit.inventory.ItemStack;
+import org.bukkit.plugin.java.JavaPlugin;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import vg.civcraft.mc.civmodcore.events.EventUtils;
+import vg.civcraft.mc.civmodcore.inventory.CustomItem;
+import vg.civcraft.mc.civmodcore.inventory.items.updater.ItemUpdater;
+import vg.civcraft.mc.civmodcore.inventory.items.updater.migrations.ItemMigrations;
+
+/**
+ * Use this in lieu of {@link vg.civcraft.mc.civmodcore.inventory.items.updater.impl.CustomItemUpdater} when you have
+ * more than one custom item that your plugin needs to migration update, since the overhead of the HashMap is likely
+ * less than the overhead of all those listeners for each and every custom item type.
+ */
+public class CustomItemsUpdater implements ItemUpdater {
+    public final NamespacedKey versionKey;
+    private final Map<String, ItemMigrations> targets;
+
+    public CustomItemsUpdater() {
+        this(ItemMigrations.DEFAULT_VERSION_KEY);
+    }
+
+    public CustomItemsUpdater(
+        final @NotNull NamespacedKey versionKey
+    ) {
+        this.versionKey = Objects.requireNonNull(versionKey);
+        this.targets = new HashMap<>();
+    }
+
+    /**
+     * Retrieves a migration list for the given custom key, creating one if it didn't already exist.
+     */
+    public @NotNull ItemMigrations getMigrationsFor(
+        final @NotNull String customKey
+    ) {
+        return getMigrationsFor(customKey, true);
+    }
+
+    /**
+     * Retrieves a migration list for the given custom key.
+     *
+     * @param createIfAbsent Whether to create the list if one doesn't already exist.
+     */
+    @Contract("_, true -> !null")
+    public @Nullable ItemMigrations getMigrationsFor(
+        final @NotNull String customKey,
+        final boolean createIfAbsent
+    ) {
+        if (createIfAbsent) {
+            return this.targets.computeIfAbsent(customKey, (_k) -> new ItemMigrations(this.versionKey));
+        }
+        return this.targets.get(customKey);
+    }
+
+    public void removeMigrationsFor(
```

Comment:
why would there be a removeMigrationsFor method? Also we don't really need two classes (overhead doesn't matter) for custom items

Reply:
> why would there be a removeMigrationsFor method?

It's there to support general reloadability. For example, SimpleAdminHacks: should there be a plugin-level `CustomItemsUpdater` that hacks use, or should each hack register its own? Hacks are designed to be as reloadable as possible. It doesn't hurt to have the method there even if you don't intend to use it.